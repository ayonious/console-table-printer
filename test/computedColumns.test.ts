import { Table } from '../index';

describe('Example: Print a simple Table with Computed Columns', () => {
  it('Computed Columns are working', () => {
    // Create a table
    const p = new Table({
      columns: [
        { name: 'red_amount', color: 'red' },
        { name: 'blue_amount', alignment: 'right' },
      ],
      computedColumns: [
        {
          name: 'sum',
          function: (row) => row.red_amount + row.blue_amount,
        },
        {
          name: 'red_percent',
          function: (row) => ((row.red_amount / row.sum) * 100).toFixed(2),
        },
        {
          name: 'blue_percent',
          function: (row) => ((row.blue_amount / row.sum) * 100).toFixed(2),
        },
      ],
    });

    // add rows with color
    p.addRow({
      red_amount: 2,
      blue_amount: 3,
    });
    p.addRow({
      red_amount: 1,
      blue_amount: 1,
    });
    p.addRow({
      red_amount: 5,
      blue_amount: 6,
    });

    // print
    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('Computed Columns are working with Title', () => {
    // Create a table
    const p = new Table({
      columns: [
        { name: 'red_amount', color: 'red' },
        { name: 'blue_amount', alignment: 'right' },
      ],
      computedColumns: [
        {
          name: 'sum',
          function: (row) => row.red_amount + row.blue_amount,
        },
        {
          name: 'red_percent',
          title: 'Red Percent',
          function: (row) => ((row.red_amount / row.sum) * 100).toFixed(2),
        },
        {
          name: 'blue_percent',
          title: 'Blue Percent',
          function: (row) => ((row.blue_amount / row.sum) * 100).toFixed(2),
        },
      ],
    });

    // add rows with color
    p.addRow({
      red_amount: 2,
      blue_amount: 3,
    });
    p.addRow({
      red_amount: 1,
      blue_amount: 1,
    });
    p.addRow({
      red_amount: 5,
      blue_amount: 6,
    });

    // print
    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('Computed Columns can calculate row numbers', () => {
    // Create a table
    const p = new Table({
      columns: [
        { name: 'red_amount', color: 'red' },
        { name: 'blue_amount', alignment: 'right' },
      ],
      computedColumns: [
        {
          name: 'row_number',
          title: 'Row Number',
          function: (row, index, rowNumbers) => index,
        },
        {
          name: 'total_rows',
          title: 'Total Rows',
          function: (row, index, rowNumbers) => rowNumbers.length,
        },
      ],
    });

    // add rows with color
    p.addRow({
      red_amount: 2,
      blue_amount: 3,
    });
    p.addRow({
      red_amount: 1,
      blue_amount: 1,
    });
    p.addRow({
      red_amount: 5,
      blue_amount: 6,
    });

    // print
    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });
});
