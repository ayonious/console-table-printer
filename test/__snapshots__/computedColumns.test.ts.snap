// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example: Print a simple Table with Computed Columns Computed Columns are working 1`] = `
"┌────────────┬─────────────┬─────┬─────────────┬──────────────┬─────┬─────────────┬──────────────┐
│[37m [0m[01mred_amount[0m │[37m [0m[01mblue_amount[0m │[37m [0m[01msum[0m │[37m [0m[01mred_percent[0m │[37m [0m[01mblue_percent[0m │[37m [0m[01msum[0m │[37m [0m[01mred_percent[0m │[37m [0m[01mblue_percent[0m │
├────────────┼─────────────┼─────┼─────────────┼──────────────┼─────┼─────────────┼──────────────┤
│[37m [0m[31m         2[0m │[37m [0m[37m          3[0m │[37m [0m[37m  5[0m │[37m [0m[37m      40.00[0m │[37m [0m[37m       60.00[0m │[37m [0m[37m  5[0m │[37m [0m[37m      40.00[0m │[37m [0m[37m       60.00[0m │
│[37m [0m[31m         1[0m │[37m [0m[37m          1[0m │[37m [0m[37m  2[0m │[37m [0m[37m      50.00[0m │[37m [0m[37m       50.00[0m │[37m [0m[37m  2[0m │[37m [0m[37m      50.00[0m │[37m [0m[37m       50.00[0m │
│[37m [0m[31m         5[0m │[37m [0m[37m          6[0m │[37m [0m[37m 11[0m │[37m [0m[37m      45.45[0m │[37m [0m[37m       54.55[0m │[37m [0m[37m 11[0m │[37m [0m[37m      45.45[0m │[37m [0m[37m       54.55[0m │
└────────────┴─────────────┴─────┴─────────────┴──────────────┴─────┴─────────────┴──────────────┘"
`;

exports[`Example: Print a simple Table with Computed Columns Computed Columns are working with Title 1`] = `
"┌────────────┬─────────────┬─────┬─────────────┬──────────────┬─────┬─────────────┬──────────────┐
│[37m [0m[01mred_amount[0m │[37m [0m[01mblue_amount[0m │[37m [0m[01msum[0m │[37m [0m[01mRed Percent[0m │[37m [0m[01mBlue Percent[0m │[37m [0m[01msum[0m │[37m [0m[01mRed Percent[0m │[37m [0m[01mBlue Percent[0m │
├────────────┼─────────────┼─────┼─────────────┼──────────────┼─────┼─────────────┼──────────────┤
│[37m [0m[31m         2[0m │[37m [0m[37m          3[0m │[37m [0m[37m  5[0m │[37m [0m[37m      40.00[0m │[37m [0m[37m       60.00[0m │[37m [0m[37m  5[0m │[37m [0m[37m      40.00[0m │[37m [0m[37m       60.00[0m │
│[37m [0m[31m         1[0m │[37m [0m[37m          1[0m │[37m [0m[37m  2[0m │[37m [0m[37m      50.00[0m │[37m [0m[37m       50.00[0m │[37m [0m[37m  2[0m │[37m [0m[37m      50.00[0m │[37m [0m[37m       50.00[0m │
│[37m [0m[31m         5[0m │[37m [0m[37m          6[0m │[37m [0m[37m 11[0m │[37m [0m[37m      45.45[0m │[37m [0m[37m       54.55[0m │[37m [0m[37m 11[0m │[37m [0m[37m      45.45[0m │[37m [0m[37m       54.55[0m │
└────────────┴─────────────┴─────┴─────────────┴──────────────┴─────┴─────────────┴──────────────┘"
`;
